docs/dev/writing-code.md

```diff
@@ -190,20 +199,270 @@ Don't use this pattern unless you're adding to a class where it's already in use
 
 TODO: document when to use future vs callback
 
+## Don't Add Features
+
+When binding an API from the aws-c libraries, don't start adding extra features.
+If you're tempted to add any "special logic" that would be valuable to the other aws-crt language bindings,
+add that logic in the underlying aws-c library so that every other language can benefit.
+
+Even for trivial things like picking nice default values, put it in the underlying aws-c library.
+(see [Use `None` for Optional Arguments](#use-none-as-the-default-value-for-optional-arguments)).
+
 # Lifetime Management
 
-TODO: Write me.
-Talk about Garbage Collector vs Reference Counting. Have pictures.
-Talk about capsules. Talk about NativeResource. Warn about cycles.
-Talk about how our tests can and cannot check for leaks.
-Talk about which classes require a `close()` function, and which don't.
+## Terminology
+
+### Strong References / Reference Counting
+
+A "strong reference" is one that keeps an object alive by incrementing its reference count.
+To "release" the reference is to decrement the object's reference count.
+When all references to an object are released, its reference count goes to zero and it gets cleaned up.
+
+In pure Python code, every variable is a strong reference to an object.
+When the variable goes away, the reference is released.
+
+In C code, reference counts on Python objects are controlled using `Py_INCREF(x)` and `Py_DECREF(x)`.
+Structs from the aws-c libraries have `_acquire(x)` and `_release(x)` calls to control their reference counts.
+We'll talk more about this [later](#reference-counting-in-c).
+
+### Reference Cycle
+
+A "reference cycle" is when a circle of strong references is created.
+Reference cycles cause memory to leak because the reference counts never get to zero.
+
+Python has a [garbage collector](https://devguide.python.org/internals/garbage-collector)
+that can detect and clean up reference cycles among normal Python objects.
+HOWEVER, any cycle involving a `Py_INCREF(x)` from C creates an undetectable cycle.
+You MUST NOT create reference cycles when designing bindings.
+
+### Capsule
+
+[PyCapsule](https://docs.python.org/3/extending/extending.html#using-capsules)
+lets us bind the lifetime of a C struct to the lifetime of a Python object.
+It's a Python object that holds a C pointer and a "destructor" function pointer.
+When Python cleans up the `PyCapsule`, the destructor function will be called.
```

Comment:
Could be useful to state that PyCapsule_New() creates a reference and if the object is returned to Python, once the object goes out of scope out there it will lose the reference causing it to call its destructor unless it's been INCREF'ed somewhere else. Or maybe that's super obvious so doesn't need mentioning?